### PR TITLE
[server] add `searchString` to `getRepositoriesForAutomatedPrebuilds` – EXP-461

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
Screenshot 2023-08-17 at 10 38 27

This PR fixes an issue with hitting rate-limits when querying repositories of BBS. 

Screenshot 2023-08-17 at 11 04 14

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at b7bb548</samp>

This pull request adds a feature to search for Bitbucket Server repositories by name, project name, or project key in the Gitpod dashboard. It modifies the `getRepos` method of the `BitbucketServerApi` class and the `getProviderRepositoriesForUser` and `getRepositoriesForAutomatedPrebuilds` methods of the `GitpodServerImpl`, `RepositoryService`, and `BitbucketServerService` classes. It also adds a `searchString` parameter to the relevant API and service methods.

</details>

## Related Issue(s)

Fixes EXP-461

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - at-searchstring</li>
	<li><b>🔗 URL</b> - <a href="https://at-searchstring.preview.gitpod-dev.com/workspaces" target="_blank">at-searchstring.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - at-searchString-gha.15552</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Review by akosyakov — APPROVED** [review]
generally code looks good, not sure how to test

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 1]
```diff
@@ -1464,7 +1464,7 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
     // Projects
     async getProviderRepositoriesForUser(
         ctx: TraceContext,
-        params: { provider: string; hints?: object },
+        params: { provider: string; hints?: object; searchString?: string },
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:1467]: good point.

File: components/server/src/bitbucket-server/bitbucket-server-api.ts [hunk 2]
```diff
@@ -280,34 +280,73 @@ export class BitbucketServerApi {
...
-                userOrToken,
-                `/repos?${permission}start=${start}`,
-            );
-            if (pageResult.values) {
-                result.push(...pageResult.values);
+        const runQuery = async (params: string) => {
+            const result: BitbucketServer.Repository[] = [];
+            let isLastPage = false;
+            let start = 0;
+            let requestsLeft = cap;
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/server/src/bitbucket-server/bitbucket-server-api.ts:314]: makes sense

**Review by geropl — APPROVED** [review]
Tested, and worked flawlessly for me. Nice! :heavy_check_mark:

**Pull request merged**